probe-rs/src/vendor/nxp/sequences/nxp_armv8m.rs

```diff
@@ -725,13 +726,83 @@ impl MIMXRT118x {
         Self(())
     }
 
-    /// Create a sequence handle for the MIMXRT117x.
+    /// Create a sequence handle for the MIMXRT118x.
     pub fn create() -> Arc<dyn ArmDebugSequence> {
         Arc::new(Self::new())
     }
+
+    /// Necessary for AIRCR.SYSRESETREQ to not be a noop
+    /// Reference:
+    /// binaries/Scripts/RT1180_connect_M33_wake_M7.scp:53 in Linkserver
+    /// `27.3.3.2` in `i.MX RT1180 Reference Manual, Rev. 6, 09/2024`
+    fn clear_cm33_reset_mask(
+        &self,
+        interface: &mut dyn ArmMemoryInterface,
+    ) -> Result<(), ArmError> {
+        const SRC_SRMASK: u64 = 0x54460018;
+        let mut srmask = interface.read_word_32(SRC_SRMASK)?;
+        tracing::trace!("SRC.SRMASK: {srmask:#010X}. Clearing the CM33_RESET_MASK mask...");
+        srmask &= !(0b1 << 8);
+        interface.write_word_32(SRC_SRMASK, srmask)?;
+        interface.flush()
+    }
 }
 
 impl ArmDebugSequence for MIMXRT118x {
+    fn reset_system(
+        &self,
+        interface: &mut dyn ArmMemoryInterface,
+        _core_type: crate::CoreType,
+        _: Option<u64>,
+    ) -> Result<(), ArmError> {
+        tracing::trace!("MIMXRT118x reset system");
+
+        self.clear_cm33_reset_mask(interface)?;
+        let mut aircr = Aircr(0);
+        aircr.vectkey();
+        aircr.set_sysresetreq(true);
+        interface.write_word_32(Aircr::get_mmio_address(), aircr.into())?;
+        interface.flush()?;
+        tracing::trace!("Reset requested..");
+        // After SYSRESETREQ, a _short_ sleep seems to be necessary. Otherwise debug interface enters some lock-up state.
+        // NXP's Linkserver sends a vendor CMSIS DAP command to MCU-Link after reset which seems to be just a ~50ms sleep.
+        // Doing the same seems to solve the issue ™️.
+        //
+        // It is pretty much the only reason why we cannot use a vanilla `cortex_m_reset_system`.
+        thread::sleep(Duration::from_millis(50));
+
+        let start = Instant::now();
+        while start.elapsed() < Duration::from_millis(5000) {
+            let dhcsr = match interface.read_word_32(Dhcsr::get_mmio_address()) {
+                Ok(val) => Dhcsr(val),
+                // Some combinations of debug probe and target (in
+                // particular, hs-probe and ATSAMD21) result in
+                // register read errors while the target is
+                // resetting.
+                Err(ArmError::AccessPort {
+                    source: crate::architecture::arm::ap::AccessPortError::RegisterRead { .. },
+                    ..
+                }) => continue,
+                Err(err) => return Err(err),
+            };
+            if dhcsr.s_reset_st() {
+                if dhcsr.s_halt() {
+                    // NXP's flash algos use the stack. Default limit on stack
+                    // makes the very first instruction (`push`) crash.
+                    // TODO: This should go the flash algo code maybe?
```

Comment:
@Yatekii what do you think about this? This seems a little hacky to adjust a MSPLIM here, especially that it is done specifically for flash algos. Maybe it's fine considering the amount of complexity we would have to introduce to the flasher as these registers are ARMv8m specific 🤔 I will remove the TODO if you think this should just live here.

Reply:
I think this would better fit in https://docs.rs/probe-rs/latest/probe_rs/architecture/arm/sequences/trait.ArmDebugSequence.html#method.debug_device_unlock or in https://docs.rs/probe-rs/latest/probe_rs/architecture/arm/sequences/trait.ArmDebugSequence.html#method.debug_core_start.
But otherwise I think this is okay.
Do you get any notion of a pack from the manufacturer or nothing at all?

Reply:
Nice thing about this is that I'm piggy-backing on the halted state here and that I'm not touching the register until I'm doing `ResetCatchSet+ResetSystem` - the typical flow when prepping flash algorithms. Otherwise, if implemented as part of `DebugCoreStart` or `DebugDeviceUnlock`, eg. `probe-rs attach` will have to unnecessarily halt the core, change MSPLIM_S and continue, right?

> any notion of a pack

Do you mean CMSIS-Pack? PDSC file contains a following debug sequence for CM33:

```xml
          <sequence Pname="cm33" name="HaltCM33">
            <block>
    </block>
          </sequence>
```

:D

NXP's LinkServer seems to use alternative flash algorithms called flash drivers which seem to set this register at some point of their execution.

I started this discussion because @korken89 thought that maybe this should be a responsibility of the flash algorithm runner/flasher/loader. Not the hill I'm willing to die on :D

Reply:
Wont any code that you try to run without this prep crash? Not just the flash algo?
If the former is true I think having it in `DebugDeviceUnlock ` e.g. would be more suitable. If this is for flash algos only (I don't see how) it should go in the algo imo. But ofc it's nice to use the one from the pack.

Reply:
I revisited the problem from the start and I was wrong. Weird MSPLIM_S values that I thought were a default were in fact a result of a broken reset catch that I fixed at some later stage. So it was a dirty left-over state of a previous application. I think we can do without any register manipulation.